src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jobs/oneoff/migration/ChangelogMigrationJobListener.kt

```diff
@@ -0,0 +1,33 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jobs.oneoff.migration
+
+import mu.KLogging
+import org.springframework.batch.core.JobExecution
+import org.springframework.batch.core.JobExecutionListener
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ChangelogRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.AmendTopic
+
+@Component
+class ChangelogMigrationJobListener(
+  val changelogRepository: ChangelogRepository
+) : JobExecutionListener {
+  companion object : KLogging()
+
+  override fun beforeJob(jobExecution: JobExecution) {
+    logger.info("Referral Details to Change log Batch Job Starts")
+    logger.info("Number of change log items before Job is = {}", changelogRepository.count())
+    val changeLogs = changelogRepository.findAll()
+    val changeLogsFilteredElements = changeLogs.filter { it.topic == AmendTopic.COMPLETION_DATETIME || it.topic == AmendTopic.MAXIMUM_ENFORCEABLE_DAYS }
+    val changelogWithNull = changeLogs.filter { changelog -> changelog.oldVal.values.contains("null") }
+    logger.info("Number of change log items with null value before Job is = {}", changelogWithNull.size)
+    logger.info("Deleting all the change log with COMPLETION_DATETIME and MAXIMUM_ENFORCEABLE_DAYS")
+    changeLogsFilteredElements.forEach { changelogRepository.deleteById(it.id) }
```

Comment:
_not a blocker, user-experience impact_

As far as I know, this deletion will remove the changelog entries in a separate transaction, so users can observe these changelogs disappearing and refilling.

Aborting the jobs for any reason (deployment/relocation of kubernetes resources) will mean the intermediate state will linger around for longer.

We had about ~400 requests to changelogs in production, so this isn't an issue in this data migration.

Later, to support updates at any time in any traffic scenario, a `process` which changes one referral in one database transaction at a time would be more suitable.

Reply:
Agreed David, In the future jobs, we will consider process using each referral